src/vos/vos_tree.c

```diff
@@ -615,58 +613,54 @@ static int
 svt_rec_fetch(struct btr_instance *tins, struct btr_record *rec,
 	       d_iov_t *key_iov, d_iov_t *val_iov)
 {
-	struct vos_key_bundle	*kbund = NULL;
+	struct vos_svt_key	*key = NULL;
 	struct vos_rec_bundle	*rbund;
 
 	rbund = iov2rec_bundle(val_iov);
 	if (key_iov != NULL)
-		kbund = iov2key_bundle(key_iov);
+		key = iov2svt_key(key_iov);
 
-	svt_rec_load(tins, rec, kbund, rbund);
+	svt_rec_load(tins, rec, key, rbund);
 	return 0;
 }
 
 static int
 svt_rec_update(struct btr_instance *tins, struct btr_record *rec,
 		d_iov_t *key_iov, d_iov_t *val_iov)
 {
-	struct svt_hkey		*skey;
-	struct vos_key_bundle	*kbund;
+	struct vos_svt_key	*skey;
+	struct vos_irec_df	*irec;
 	struct vos_rec_bundle	*rbund;
+	int			 rc;
 
-	kbund = iov2key_bundle(key_iov);
 	rbund = iov2rec_bundle(val_iov);
+	skey = (struct vos_svt_key *)key_iov->iov_buf;
+	irec = vos_rec2irec(tins, rec);
 
-	if (!UMOFF_IS_NULL(rbund->rb_off) ||
-	    !vos_irec_size_equal(vos_rec2irec(tins, rec), rbund)) {
-		/* This function should return -DER_NO_PERM to dbtree if:
-		 * - it is a rdma, the original record should be replaced.
-		 * - the new record size cannot match the original one, so we
-		 *   need to realloc and copyin data to the new space.
-		 *
-		 * So dbtree can release the original record and install the
-		 * rdma-ed record, or just allocate a new one.
-		 */
+	/** Disallow same epoch overwrite */
+	if (skey->sk_minor_epc <= irec->ir_minor_epc)
 		return -DER_NO_PERM;
-	}
 
-	skey = (struct svt_hkey *)&rec->rec_hkey[0];
-	D_DEBUG(DB_IO, "Overwrite epoch "DF_U64"\n", skey->sv_epoch);
+	D_DEBUG(DB_IO, "Overwrite epoch "DF_X64".%d\n", skey->sk_epoch,
+		skey->sk_minor_epc);
+
+	rc = svt_rec_free(tins, rec, NULL);
```

Comment:
If the value size is the same, can we take over the original space and overwrite directly?

Reply:
Sure but the new one is already allocated at this point anyway so it doesn't really matter much